main.cpp

```diff
@@ -322,11 +390,61 @@ int uthread_terminate(int tid) {
     } else if((it = find(waiting_list.begin(), waiting_list.end(), tid)) != waiting_list.end()) {
 	waiting_list.erase(it);
     } else {
-	return -1;
+	enable_interrupts();
+	thread_complete();  // If tid is the running thread
+	return 0;
     }
 
     free_waiting_threads(tid);
+
+    enable_interrupts();
+    return 0;
+}
+
+/**
+ * Sets the time slice for how long each thread runs
+ * @param time_slice - the new time slice for each thread in microseconds 
+ */
+int uthread_init(int time_slice) {
+    timer.it_interval.tv_usec = time_slice;
+    return 0;
+}
+
+/**
+ * Calls thread_yield every time the timer expires
+ */
+void timer_handler(int signum) {
+    static int count = 0;
+    printf("Timer expired: %d\n", ++count);
+    uthread_yield();
```

Comment:
I am not sure that this is going to work. See https://en.wikipedia.org/wiki/C_signal_handling#Handling. Usually, you are supposed to return from the signal handler if you don't close the program. I tested on https://gist.github.com/ef5a4d2c0ed009ca282b0573a76eb15f, and it doesn't appear to work - you only get the signal the first time.

Reply:
I think that non-local jumps are okay in the handler. I don't know how else it's possible to implement the time slicing. 

I think the test program you wrote needs to use "sigsetjump/siglongjmp" instead of just "setjmp/longjmp" in order to work.

Reply:
https://www.gnu.org/software/libc/manual/html_node/Longjmp-in-Handler.html#Longjmp-in-Handler

Reply:
Yeah, you're right.